tlsn-core/src/utils.rs

```diff
@@ -0,0 +1,266 @@
+fn merged_slice(
+    a: &TranscriptSlice,
+    b: &TranscriptSlice,
+) -> Result<Option<TranscriptSlice>, Error> {
+    let merged_range = match merged_range(a.range(), b.range()) {
+        // ranges neither overlap nor are adjacent
+        None => return Ok(None),
+        Some(range) => range,
+    };
+
+    // `merged_data` will contain data of the merged output slice
+    let mut merged_data: Vec<u8> = vec![0u8; merged_range.len() as usize];
+
+    // copy data from both slices into the output slice, ignoring for now the fact that
+    // overlapping data will be overwritten
+    let a_start_offset = (a.range().start() - merged_range.start()) as usize;
+    merged_data[a_start_offset..a_start_offset + a.data().len()].copy_from_slice(a.data());
+
+    let b_start_offset = (b.range().start() - merged_range.start()) as usize;
+    merged_data[b_start_offset..b_start_offset + b.data().len()].copy_from_slice(b.data());
+
+    // by checking that both a and b are subsets of the output slice, we make sure that a and b have
+    // the exact same data in the overlapping range
+    if &merged_data[a_start_offset..a_start_offset + a.data().len()] != a.data()
+        || &merged_data[b_start_offset..b_start_offset + b.data().len()] != b.data()
+    {
+        return Err(Error::OverlappingSlicesDontMatch);
+    }
+
+    Ok(Some(TranscriptSlice::new(merged_range, merged_data)))
+}
+
+/// If two ranges overlap or are adjacent, returns a merged range
+fn merged_range(a: &TranscriptRange, b: &TranscriptRange) -> Option<TranscriptRange> {
+    if !is_overlapping_or_adjacent(a, b) {
+        return None;
+    }
+    // overlap detected
+    let merged_start = std::cmp::min(a.start(), b.start());
+    let merged_end = std::cmp::max(a.end(), b.end());
+
+    Some(TranscriptRange::new(merged_start, merged_end).expect("start bound must be > end bound"))
+}
+
+/// Returns true if two ranges overlap or are adjacent
+fn is_overlapping_or_adjacent(a: &TranscriptRange, b: &TranscriptRange) -> bool {
+    // find purported overlap's start and end
+    let ov_start = std::cmp::max(a.start(), b.start());
+    let ov_end = std::cmp::min(a.end(), b.end());
+
+    // Note that even if ranges do not overlap, they may still be adjacent if
+    // ov_start == ov_end
+    // (prevent overflow panic by casting into i64)
+    ov_end as i64 - ov_start as i64 >= 0
+}
+
+/// Encodes the bytes located in ranges and returns a vec of encodings
+/// It is only called internally when we know that
+/// A) bytes.len() is equal to the sum of lengths of all ranges
+/// B) ranges are non-overlapping and in ascending order
+pub(crate) fn encode_bytes_in_ranges(
+    encoder: &ChaChaEncoder,
+    bytes: &[u8],
+    ranges: &[TranscriptRange],
+    direction: &Direction,
+) -> Vec<[u8; 16]> {
+    // dummy id. In reality, the id will be generated differently
+    let id = if direction == &Direction::Sent { 0 } else { 1 };
```

Comment:
What exactly do you mean by that?

Reply:
in reality we will have a unique encoding id for each byte in the transcript, but here i used the same id for all bytes in all ranges. 
But this method was only for demo purposes until I integrate with the mpc encoder